---

title: GTEx v8 - Download and Basic library preparation


keywords: fastai
sidebar: home_sidebar



nb_path: "01_GTEx.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 01_GTEx.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210128 16:40:25 gtex:35] Found /home/dustin/fleet_gene/data
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDD.xlsx
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt
[I 210128 16:40:25 gtex:42] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SampleAttributesDD.xlsx
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">glob</span>

<span class="kn">import</span> <span class="nn">inspect</span><span class="o">,</span> <span class="nn">urllib</span>

<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>

<span class="kn">import</span> <span class="nn">time</span><span class="o">,</span> <span class="nn">datetime</span>

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="kn">import</span> <span class="nn">plotly.express</span> <span class="k">as</span> <span class="nn">px</span>

<span class="c1"># from chart_studio import plotly as py</span>
<span class="kn">import</span> <span class="nn">plotly.figure_factory</span> <span class="k">as</span> <span class="nn">ff</span>
<span class="kn">import</span> <span class="nn">plotly.graph_objects</span> <span class="k">as</span> <span class="nn">go</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Download-GTEX-data-locally">Download GTEX data locally<a class="anchor-link" href="#Download-GTEX-data-locally"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># Detailed data file:</span>
<span class="n">gtex_tpm_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEXV8_TPM</span><span class="p">))</span>

<span class="c1"># Just looking at median values by tissue (SMTSD) greatly reduces file size</span>
<span class="n">gtex_tpm_med_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEXV8_TPM_MED</span><span class="p">))</span>


<span class="c1"># Sample annotation data</span>

<span class="c1"># Subject Phenotype data</span>
<span class="c1"># Age, sex and Hardy Scale death circumstances</span>
<span class="n">gtex_pheno_fn</span> <span class="o">=</span>  <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEX_PHENO_DS</span><span class="p">))</span>

<span class="c1"># Main sample data of interest is:</span>
<span class="c1"># &#39;SMTS&#39;: &#39;Tissue Type, area from which the tissue sample was taken.  This is a parent value to SMTSD.&#39;</span>
<span class="c1"># &#39;SMTSD&#39;: &#39;SMTS Detailed&#39;</span>
<span class="n">gtex_attr_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEX_SAMPLE_DS</span><span class="p">))</span>
<span class="n">gtex_attr_desc_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEX_SAMPLE_DD</span><span class="p">))</span>

<span class="c1"># Load tissue specific detatils</span>
<span class="n">gtex_attr</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gtex_attr_fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">tissue_types</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTS&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="n">tissue_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="n">tissue_types</span><span class="p">:</span>
    <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTS&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;SAMPID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">subtype</span> <span class="ow">in</span> <span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTS&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;SMTSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="c1"># Get all samples by subtype</span>
        <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">][</span><span class="n">subtype</span><span class="p">]</span> <span class="o">=</span> <span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTSD&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">subtype</span><span class="p">][</span><span class="s1">&#39;SAMPID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    

    <span class="nb">sorted</span><span class="p">(</span><span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTS&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;SMTSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="n">tissue_subtypes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SMTSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

<span class="c1"># log number of tissue samples</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> separate tissue types - total samples </span><span class="si">%s</span><span class="s1">&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s1">&#39;SAMPID&#39;</span><span class="p">]))</span>
<span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tissue_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> (n=</span><span class="si">%s</span><span class="s1">)&#39;</span><span class="p">,</span> <span class="n">tissue</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">]))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">subtype</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="si">%s</span><span class="s1"> (n=</span><span class="si">%s</span><span class="s1">)&#39;</span><span class="p">,</span> <span class="n">subtype</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">][</span><span class="n">subtype</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:39] 31 separate tissue types - total samples 22951
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Adipose Tissue (n=1327)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Adipose - Subcutaneous (n=763)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Adipose - Visceral (Omentum) (n=564)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Adrenal Gland (n=275)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Bladder (n=21)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Blood (n=3480)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Cells - EBV-transformed lymphocytes (n=192)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Whole Blood (n=3288)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Blood Vessel (n=1473)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Artery - Aorta (n=450)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Artery - Coronary (n=253)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Artery - Tibial (n=770)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Bone Marrow (n=217)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Brain (n=3326)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Amygdala (n=177)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Anterior cingulate cortex (BA24) (n=213)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Caudate (basal ganglia) (n=291)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Cerebellar Hemisphere (n=263)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Cerebellum (n=298)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Cortex (n=325)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Frontal Cortex (BA9) (n=425)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Hippocampus (n=243)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Hypothalamus (n=236)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Nucleus accumbens (basal ganglia) (n=277)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Putamen (basal ganglia) (n=232)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Spinal cord (cervical c-1) (n=182)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Brain - Substantia nigra (n=164)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Breast (n=480)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Cervix Uteri (n=19)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Cervix - Ectocervix (n=9)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Cervix - Endocervix (n=10)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Colon (n=821)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Colon - Sigmoid (n=389)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Colon - Transverse (n=432)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Esophagus (n=1582)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Esophagus - Gastroesophageal Junction (n=401)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Esophagus - Mucosa (n=622)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Esophagus - Muscularis (n=559)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Fallopian Tube (n=9)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Heart (n=1141)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Heart - Atrial Appendage (n=452)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Heart - Left Ventricle (n=689)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Kidney (n=104)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Kidney - Cortex (n=100)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Kidney - Medulla (n=4)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Liver (n=251)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Lung (n=867)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Muscle (n=1132)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Nerve (n=722)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Ovary (n=195)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Pancreas (n=360)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Pituitary (n=301)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Prostate (n=262)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Salivary Gland (n=181)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Skin (n=2014)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Cells - Cultured fibroblasts (n=527)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Skin - Not Sun Exposed (Suprapubic) (n=638)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:44] 	Skin - Sun Exposed (Lower leg) (n=849)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Small Intestine (n=193)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Spleen (n=260)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Stomach (n=381)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Testis (n=406)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Thyroid (n=812)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Uterus (n=166)
[I 210125 10:35:01 &lt;ipython-input-3-50b7829108bd&gt;:41] Vagina (n=173)
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">):</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">cache_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">)</span>

<span class="k">for</span> <span class="n">url</span> <span class="ow">in</span> <span class="p">[</span><span class="n">GTEXV8_TPM_MED</span><span class="p">,</span> <span class="n">GTEXV8_TPM</span><span class="p">,</span> <span class="n">GTEX_PHENO_DS</span><span class="p">,</span> <span class="n">GTEX_PHENO_DD</span><span class="p">,</span> <span class="n">GTEX_SAMPLE_DS</span><span class="p">,</span> <span class="n">GTEX_SAMPLE_DD</span><span class="p">]:</span>
    <span class="n">dest</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">url</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">dest</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;found existing: </span><span class="si">{</span><span class="n">dest</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Downloading </span><span class="si">{</span><span class="n">dest</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">stream</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dest</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">iter_content</span><span class="p">()):</span>
                <span class="n">fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Completed </span><span class="si">{</span><span class="n">dest</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:4] Found /home/dustin/fleet_gene/data
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDD.xlsx
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt
[I 210125 12:40:00 &lt;ipython-input-27-16ead20aa256&gt;:11] found existing: /home/dustin/fleet_gene/data/GTEx_Analysis_v8_Annotations_SampleAttributesDD.xlsx
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># The full set of all all GTEx tpms is very large and can be hard to work with</span>
<span class="c1"># Some functions only load some of the data at a time.</span>

<span class="c1"># Genes of interest - some test genes </span>
<span class="n">TEST_GENES</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;TP53&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ERBB2&#39;</span><span class="p">,</span>  <span class="c1"># Herceptin target</span>
    <span class="s1">&#39;EGFR&#39;</span><span class="p">,</span>
    <span class="s1">&#39;AKT1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;KRAS&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span>
    <span class="s1">&#39;APOE&#39;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">TUMOUR_MUT_GENES</span> <span class="o">=</span> <span class="p">[</span>  <span class="c1"># www.tumourportal.org - highly mutated</span>
    <span class="s1">&#39;TP53&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PIK3CA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span>
    <span class="s1">&#39;KRAS&#39;</span><span class="p">,</span>
    <span class="s1">&#39;APC&#39;</span><span class="p">,</span>
    <span class="s1">&#39;MLL3&#39;</span><span class="p">,</span>  <span class="c1"># aka KMT2C</span>
    <span class="s1">&#39;KMT2C&#39;</span><span class="p">,</span>
    <span class="s1">&#39;FAT1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;MLL2&#39;</span><span class="p">,</span>  <span class="c1"># aka KMT2D</span>
    <span class="s1">&#39;KMT2D&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ARID1A&#39;</span><span class="p">,</span>
    <span class="s1">&#39;VHL&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PBRM1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;NF1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;EGFR&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ATM&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PIK3R1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;BRAF&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CDKN2A&#39;</span><span class="p">,</span>
    <span class="s1">&#39;SETD2&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CREBBP&#39;</span><span class="p">,</span>
    <span class="s1">&#39;FBXW7&#39;</span><span class="p">,</span>
    <span class="s1">&#39;SPEN&#39;</span><span class="p">,</span>
    <span class="s1">&#39;MTOR&#39;</span><span class="p">,</span>
    <span class="s1">&#39;RB1&#39;</span><span class="p">,</span>
    <span class="s1">&#39;SMARCA4&#39;</span><span class="p">,</span>
    <span class="s1">&#39;NOTCH1&#39;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">TEST_GENES</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">TEST_GENES</span> <span class="o">+</span> <span class="n">TUMOUR_MUT_GENES</span><span class="p">))</span>

<span class="c1"># gtex_tpm file has genes as rows.</span>
<span class="c1"># First two columns are gene information</span>
<span class="c1"># Over 10000 columns of samples</span>

<span class="c1">#                Name Description  GTEX-1117F-0226-SM-5GZZ7  GTEX-111CU-1826-SM-5GZYN ...</span>
<span class="c1">#0  ENSG00000223972.4     DDX11L1                   0.10820                   0.11580</span>
<span class="c1">#1  ENSG00000227232.4      WASH7P                  21.40000                  11.03000</span>


<span class="n">GTPM_SKIP</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1">#  skip two comment rows at the beginning of the file</span>

<span class="c1"># Just getting a few rows of this giant file</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Loading Row partial table ...&#39;</span><span class="p">)</span>
<span class="n">gtpm_row_table</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gtex_tpm_fn</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">))</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Loading column partial table ...&#39;</span><span class="p">)</span>
<span class="n">gtpm_column_table</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gtex_tpm_fn</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">gtpm_row_by_genename</span><span class="p">(</span><span class="n">gene_name</span><span class="p">):</span>
    <span class="n">sub_table</span> <span class="o">=</span> <span class="n">gtpm_row_table</span><span class="p">[</span><span class="n">gtpm_row_table</span><span class="p">[</span><span class="s1">&#39;Description&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene_name</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">sub_table</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Missing &#39;</span><span class="si">%s</span><span class="s2"> row in GTEX TPM table&#39;&quot;</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">GTPM_SKIP</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sub_table</span><span class="o">.</span><span class="n">index</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

<span class="k">def</span> <span class="nf">gtpm_keep_rows</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">TEST_GENES</span><span class="p">):</span>
    <span class="n">row_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">gtpm_row_by_genename</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">GTPM_SKIP</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">row_list</span> <span class="k">if</span> <span class="n">gene</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">gtex_tpm_partial_table_load</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">sample_list</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">gene_list</span><span class="p">:</span>
        <span class="n">keep_rows</span> <span class="o">=</span> <span class="n">gtpm_keep_rows</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>
        <span class="n">skiprows_func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">keep_rows</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;loading </span><span class="si">%s</span><span class="s1"> GTEX TPM rows&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">keep_rows</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;loading all rows&#39;</span><span class="p">)</span>
        <span class="n">skiprows_func</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="n">sample_list</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Limiting load to </span><span class="si">%s</span><span class="s1"> potential samples&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_list</span><span class="p">))</span>
        <span class="c1"># Not all samples are in tables</span>
        <span class="n">sample_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">sample_list</span> <span class="k">if</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">gtpm_column_table</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Limiting load to </span><span class="si">%s</span><span class="s1"> samples&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_list</span><span class="p">))</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Name&#39;</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">sample_list</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sample_list</span> <span class="o">=</span> <span class="kc">None</span>
    
    <span class="n">gtpm_partial</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">gtex_tpm_fn</span><span class="p">,</span>
        <span class="n">skiprows</span><span class="o">=</span><span class="n">skiprows_func</span><span class="p">,</span>  <span class="c1"># Only selected gene rows</span>
        <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">usecols</span><span class="o">=</span><span class="n">sample_list</span><span class="p">,</span>  <span class="c1"># Number of samples limit</span>
    <span class="p">)</span>
    <span class="n">end_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Gtex TPM partial table took </span><span class="si">{}</span><span class="s2"> minutes&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">((</span><span class="n">end_time</span> <span class="o">-</span> <span class="n">start_time</span><span class="p">)</span><span class="o">/</span><span class="mi">60</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">gtpm_partial</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210125 10:54:13 &lt;ipython-input-4-b213538e8d2c&gt;:61] Loading Row partial table ...
[I 210125 10:54:47 &lt;ipython-input-4-b213538e8d2c&gt;:63] Loading column partial table ...
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Cache-sample-and-gene-lists">Cache sample and gene lists<a class="anchor-link" href="#Cache-sample-and-gene-lists"> </a></h3><p>Preload and save these lists so that we can load part of the GTEX tables easily</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">gtpm_column_table</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">3</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">gtpm_column_table</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Name&quot;</span>
<span class="k">assert</span> <span class="n">gtpm_column_table</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Description&quot;</span>

<span class="c1"># Remaining columns are GTEX samples that start with GTEX-</span>
<span class="n">GTEX_SAMPLES</span> <span class="o">=</span> <span class="n">gtpm_column_table</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">GTEX_SAMPLES</span><span class="p">:</span>
    <span class="k">assert</span> <span class="n">sample</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GTEX-&quot;</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GTEX data has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">GTEX_SAMPLES</span><span class="p">)</span><span class="si">}</span><span class="s2"> samples&quot;</span><span class="p">)</span>

<span class="n">GTEX_GENES_ENSG</span> <span class="o">=</span> <span class="n">gtpm_row_table</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">GTEX_GENES_SYMBOL</span> <span class="o">=</span> <span class="n">gtpm_row_table</span><span class="p">[</span><span class="s2">&quot;Description&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">GTEX_GENES_SYMBOL</span><span class="p">[:</span><span class="mi">2</span><span class="p">])</span>

<span class="n">pseudo</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">ensg</span><span class="p">,</span> <span class="n">symbol</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">,</span> <span class="n">GTEX_GENES_SYMBOL</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">ensg</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENSG&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="s2">&quot;.&quot;</span> <span class="ow">in</span> <span class="n">ensg</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">ensg</span><span class="p">[</span><span class="n">ensg</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="nb">int</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">version</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;_PAR_Y&quot;</span><span class="p">):</span>
            <span class="c1"># PAR1, PAR2, are homologous sequences of nucleotides on the X and Y chromosomes</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nonstandard ENSG in </span><span class="si">{</span><span class="n">symbol</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">ensg</span><span class="si">}</span><span class="s2"> version: </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">pseudo</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">ensg</span><span class="p">,</span> <span class="n">symbol</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Nonstandard ENSG </span><span class="si">{</span><span class="n">ensg</span><span class="si">}</span><span class="s2"> version: </span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GTEX data on </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:10] GTEX data has 17382 samples
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in LL0YNC03-29C1.1 ENSG00000228572.7_PAR_Y version: 7_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in PLCXD1 ENSG00000182378.13_PAR_Y version: 13_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in GTPBP6 ENSG00000178605.13_PAR_Y version: 13_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in LINC00685 ENSG00000226179.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in PPP2R3B ENSG00000167393.17_PAR_Y version: 17_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RP13-465B17.4 ENSG00000281849.3_PAR_Y version: 3_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in Metazoa_SRP ENSG00000275287.5_PAR_Y version: 5_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RP13-465B17.5 ENSG00000280767.3_PAR_Y version: 3_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in FABP5P13 ENSG00000234958.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in KRT18P53 ENSG00000229232.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in SHOX ENSG00000185960.13_PAR_Y version: 13_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RP11-309M23.1 ENSG00000237531.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RPL14P5 ENSG00000225661.7_PAR_Y version: 7_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in CRLF2 ENSG00000205755.11_PAR_Y version: 11_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in CSF2RA ENSG00000198223.16_PAR_Y version: 16_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in MIR3690 ENSG00000265658.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RNA5SP498 ENSG00000223274.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in IL3RA ENSG00000185291.11_PAR_Y version: 11_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in SLC25A6 ENSG00000169100.13_PAR_Y version: 13_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in LINC00106 ENSG00000236871.7_PAR_Y version: 7_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in ASMTL-AS1 ENSG00000236017.8_PAR_Y version: 8_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in ASMTL ENSG00000169093.15_PAR_Y version: 15_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in P2RY8 ENSG00000182162.10_PAR_Y version: 10_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in AKAP17A ENSG00000197976.11_PAR_Y version: 11_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in ASMT ENSG00000196433.12_PAR_Y version: 12_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RP13-297E16.4 ENSG00000223511.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in RP13-297E16.5 ENSG00000234622.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in DHRSX ENSG00000169084.13_PAR_Y version: 13_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in DHRSX-IT1 ENSG00000223571.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in ZBED1 ENSG00000214717.11_PAR_Y version: 11_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in MIR6089 ENSG00000277120.5_PAR_Y version: 5_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in CD99P1 ENSG00000223773.7_PAR_Y version: 7_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in LINC00102 ENSG00000230542.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in CD99 ENSG00000002586.18_PAR_Y version: 18_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in SPRY3 ENSG00000168939.11_PAR_Y version: 11_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in AMD1P2 ENSG00000237801.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in DPH3P2 ENSG00000237040.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in VAMP7 ENSG00000124333.15_PAR_Y version: 15_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in ELOCP24 ENSG00000228410.6_PAR_Y version: 6_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in TRPC6P ENSG00000223484.7_PAR_Y version: 7_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in IL9R ENSG00000124334.17_PAR_Y version: 17_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in WASIR1 ENSG00000185203.12_PAR_Y version: 12_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in WASH6P ENSG00000182484.15_PAR_Y version: 15_PAR_Y
[D 210125 12:17:52 &lt;ipython-input-19-ce4b51075d0b&gt;:27] Nonstandard ENSG in DDX11L16 ENSG00000227159.8_PAR_Y version: 8_PAR_Y
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Index([&#39;Name&#39;, &#39;Description&#39;, &#39;GTEX-1117F-0226-SM-5GZZ7&#39;], dtype=&#39;object&#39;)
[&#39;ENSG00000223972.5&#39;, &#39;ENSG00000227232.5&#39;]
[&#39;DDX11L1&#39;, &#39;WASH7P&#39;]
GTEX data on 56200
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">dup_symbols</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">GTEX_GENES_SYMBOL</span> <span class="k">if</span> <span class="n">GTEX_GENES_SYMBOL</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">num_ambiguous_recs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">dup_symbols</span><span class="p">)</span>
<span class="n">dup_symbols</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">dup_symbols</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GTEX has </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">dup_symbols</span><span class="p">)</span><span class="si">}</span><span class="s2"> duplicate &#39;Descriptions&#39; involving </span><span class="si">{</span><span class="n">num_ambiguous_recs</span><span class="si">}</span><span class="s2"> Ensembl genes.&quot;</span><span class="p">)</span>

<span class="c1"># Verify there are only duplications in symbols and not Ensembl gene ids</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">))</span>

<span class="c1"># pickle these lists for quick reloading later</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Pickling gene lists to </span><span class="si">{</span><span class="n">cache_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">GTEX_GENES_ENSG</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="s2">&quot;GTEX_GENES_ENSG.pkl&quot;</span><span class="p">),</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
<span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">GTEX_GENES_SYMBOL</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="s2">&quot;GTEX_GENES_SYMBOL.pkl&quot;</span><span class="p">),</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
<span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">dup_symbols</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="s2">&quot;GTEX_GENES_SYMBOL_DUPLICATES.pkl&quot;</span><span class="p">),</span> <span class="s2">&quot;wb&quot;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>[I 210125 12:41:48 &lt;ipython-input-28-9f21507b2af4&gt;:10] Pickling gene lists to /home/dustin/fleet_gene/data
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>GTEX has 199 duplicate &#39;Descriptions&#39; involving 1807 Ensembl genes.
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;loading median GTEx tissue TPM values: </span><span class="si">{</span><span class="n">gtex_tpm_med_fn</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">gtpm_med</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gtex_tpm_med_fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># gtpm_breast = gtex_tpm_partial_table_load(gene_list=TEST_GENES, sample_list=tissue_dict[&#39;Breast&#39;][&#39;samples&#39;])</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;loading GTEX TPMs for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">TEST_GENES</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot; GTEX GENES </span><span class="si">{</span><span class="n">TEST_GENES</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">gtpm</span> <span class="o">=</span> <span class="n">gtex_tpm_partial_table_load</span><span class="p">(</span><span class="n">gene_list</span><span class="o">=</span><span class="n">TEST_GENES</span><span class="p">,)</span>

<span class="c1"># Remove missing genes from the test</span>
<span class="n">TEST_GENES</span> <span class="o">=</span> <span class="n">gtpm</span><span class="p">[</span><span class="s1">&#39;Description&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Median-TPM-values-and-Tissue-comparison">Median TPM values and Tissue comparison<a class="anchor-link" href="#Median-TPM-values-and-Tissue-comparison"> </a></h2><p>The smaller table of median TPM values can give us a good idea of the overall character of the different genes.</p>
<p>The spread includes extreme values, most of the tissues are similar.</p>
<p>Mean is ~ 16 - 17 TPM, but 50th percentile is 0 and 75th percentile is ~ 2.</p>
<p>This indicates that the TPM is not a normal distribution, but has a small number of extreme values.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># Do something with the median values</span>

<span class="n">rename_cols</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tissue_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="c1"># pprint(tissue_dict[t][&#39;subtypes&#39;].keys())</span>
    <span class="c1"># pick a single representative tissue subtype</span>
    <span class="n">subtype</span> <span class="o">=</span>  <span class="nb">sorted</span><span class="p">(</span><span class="n">tissue_dict</span><span class="p">[</span><span class="n">t</span><span class="p">][</span><span class="s1">&#39;subtypes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">subtype</span> <span class="o">!=</span> <span class="s1">&#39;Cells - Leukemia cell line (CML)&#39;</span><span class="p">:</span>
        <span class="n">rename_cols</span><span class="p">[</span><span class="n">subtype</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span>
<span class="n">sample_columns</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">t</span>
     <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">tissue_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">)</span>

<span class="c1"># gm = gtpm_med.drop(&#39;gene_id&#39;, axis=1).set_index(&#39;Description&#39;)</span>
<span class="n">gm</span> <span class="o">=</span> <span class="n">gtpm_med</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Description&#39;</span><span class="p">)</span>

<span class="n">gm</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">rename_cols</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">gm</span> <span class="o">=</span> <span class="n">gm</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">rename_cols</span><span class="o">.</span><span class="n">values</span><span class="p">())]</span>
<span class="c1"># Copy table so we can add and modify values</span>
<span class="n">gm</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Using gm table we can chacterize the genes a bit.</span>
<span class="c1"># the spread includes extreme values, most of the tissues are similar</span>
<span class="c1"># Mean is ~ 16 - 17 TPM, but 50th percentile is 0 and 75th percentile is ~ 2</span>
<span class="n">pandas</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;max_columns&#39;</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
<span class="n">pandas</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">120</span><span class="p">)</span>
<span class="c1">#print(gm.describe())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gm</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">TEST_GENES</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># Apply some known values</span>
<span class="n">gm</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gm</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gm</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">std</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1">#gm[&#39;mode&#39;] = gm.apply(lambda row: row.mode(), axis=1)</span>
<span class="n">gm</span><span class="p">[</span><span class="s1">&#39;CoV&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gm</span><span class="p">[</span><span class="s1">&#39;sem&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gm</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">sem</span><span class="p">(),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">gm</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>

<span class="c1"># high expression # low deviation among tissues</span>
<span class="n">gm_bench</span> <span class="o">=</span> <span class="n">gm</span><span class="p">[(</span><span class="n">gm</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># high expression</span>
             <span class="o">&amp;</span> <span class="p">(</span><span class="n">gm</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="mf">3.5</span> <span class="o">*</span> <span class="n">gm</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
            <span class="p">]</span>

<span class="c1">#cross_tissue_genes = sorted(gm_bench[&#39;Description&#39;].tolist())</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Of </span><span class="si">{}</span><span class="s2"> genes, the top </span><span class="si">{}</span><span class="s2"> in stable cross-tissue expression are:&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gm</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">gm_bench</span><span class="p">)))</span>
<span class="c1">#for gene in cross_tissue_genes:</span>
<span class="c1">#    print(&#39;\t{}&#39;.format(gene))</span>
<span class="n">gm_bench</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;CoV&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gm</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;sem&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">30</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gm</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">TEST_GENES</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gt</span> <span class="o">=</span> <span class="n">gtpm</span><span class="p">[</span><span class="n">gtpm</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]]</span>
<span class="n">gt</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Description&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">org_cols</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">columns</span>

<span class="c1">#gtt = gt.transpose()</span>
<span class="c1">#gtt.head()</span>
<span class="n">gt</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">describe</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;ERBB2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">describe</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;PTEN&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">describe</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#gt.loc[&#39;PTEN&#39;].skew()</span>
<span class="c1">#gt.loc[&#39;PTEN&#39;].kurtosis()</span>
<span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;PTEN&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="mf">0.9</span><span class="p">)</span>

<span class="n">gtpm</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="c1">#gt.T[TEST_GENES[:-1]].iplot(kind=&#39;box&#39;,)</span>
<span class="c1">#gt.T[[&#39;TP53&#39;, &#39;PTEN&#39;]].plot(kind=&#39;box&#39;,)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">math</span>

<span class="c1">#gene = &#39;TP53&#39;</span>
<span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;PTEN&#39;</span>
<span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">: Samples </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]))</span>
<span class="c1">#gtt[gene].iplot(kind=&#39;hist&#39;, title=title, yTitle=&#39;counts&#39;, xTitle=&#39;Expression (TPM)&#39;, bins=100)</span>

<span class="c1">#py.iplot(gt.loc[gene].to_list(), kind=&#39;hist&#39;, title=title, yTitle=&#39;counts&#39;, xTitle=&#39;Expression (TPM)&#39;, bins=100)</span>
<span class="c1">#gt.loc[gene].iplot(kind=&#39;hist&#39;, title=title, yTitle=&#39;counts&#39;, xTitle=&#39;Expression (TPM)&#39;, bins=100)</span>
<span class="c1">#gt.loc[gene].plot(kind=&#39;hist&#39;, title=title, bins=100)</span>

<span class="n">glog2</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">log2</span><span class="p">)</span>
<span class="n">glog2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;hist&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">px</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#print(ff.create_distplot.__doc__)</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">,</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;KRAS&#39;</span><span class="p">,</span> <span class="s1">&#39;ERBB2&#39;</span><span class="p">,</span> <span class="s1">&#39;CDKN2A&#39;</span><span class="p">]</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">,</span> <span class="s1">&#39;PTEN&#39;</span><span class="p">,</span> <span class="s1">&#39;KRAS&#39;</span><span class="p">,]</span>
<span class="n">px</span><span class="o">.</span><span class="n">histogram</span><span class="p">([</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">g</span><span class="p">]</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">],</span> <span class="n">gene_list</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;TP53&#39;</span>
<span class="n">data_lists</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="n">tissue_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span>
    <span class="c1"># print(tissue)</span>
    <span class="c1"># print(len(samples))</span>
    <span class="n">gt_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples</span> <span class="k">if</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">gt</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gt_samples</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">:</span>
        <span class="n">dt</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="n">gt_samples</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">dt</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">data_lists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt</span><span class="p">)</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">(</span><span class="si">{}</span><span class="s1">) ~ </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tissue</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">dt</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="mi">1</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only </span><span class="si">{}</span><span class="s1"> samples for &quot;</span><span class="si">{}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gt_samples</span><span class="p">),</span> <span class="n">tissue</span><span class="p">))</span>
<span class="c1"># add all samples</span>
<span class="n">dt</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">data_lists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt</span><span class="p">)</span>
<span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> - all (</span><span class="si">{}</span><span class="s1">) ~ </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dt</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">dt</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="mi">1</span><span class="p">)))</span>

<span class="n">fig_tp53</span> <span class="o">=</span> <span class="n">ff</span><span class="o">.</span><span class="n">create_distplot</span><span class="p">(</span><span class="n">data_lists</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">show_hist</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_rug</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_curve</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># curve_type=&#39;normal&#39;)</span>
<span class="n">fig_tp53</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
<span class="n">fig_tp53</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">dir</span><span class="p">(</span><span class="n">go</span><span class="p">)</span>
<span class="n">gene</span><span class="o">=</span><span class="s1">&#39;PTEN&#39;</span>
<span class="n">tpm_label</span> <span class="o">=</span> <span class="s2">&quot;TPM      </span><span class="si">{}</span><span class="s2"> skew: </span><span class="si">{}</span><span class="s2"> kurtosis: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span>
                                                       <span class="nb">round</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">skew</span><span class="p">(),</span> <span class="mi">2</span><span class="p">),</span>
                                                       <span class="nb">round</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">kurtosis</span><span class="p">(),</span> <span class="mi">2</span><span class="p">))</span>
<span class="n">log2_label</span> <span class="o">=</span> <span class="s2">&quot;log2TPM </span><span class="si">{}</span><span class="s2"> skew: </span><span class="si">{}</span><span class="s2"> kurtosis: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span>
                                                       <span class="nb">round</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">)</span><span class="o">.</span><span class="n">skew</span><span class="p">(),</span> <span class="mi">2</span><span class="p">),</span>
                                                       <span class="nb">round</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">)</span><span class="o">.</span><span class="n">kurtosis</span><span class="p">(),</span> <span class="mi">2</span><span class="p">))</span>

<span class="n">fig_log2vsNormal</span> <span class="o">=</span> <span class="n">ff</span><span class="o">.</span><span class="n">create_distplot</span><span class="p">([</span><span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">),</span> <span class="n">gt</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]],</span>
                         <span class="p">[</span><span class="n">log2_label</span><span class="p">,</span> <span class="n">tpm_label</span><span class="p">],</span>
                         <span class="n">bin_size</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span>
                         <span class="n">curve_type</span><span class="o">=</span><span class="s1">&#39;normal&#39;</span><span class="p">,)</span>
<span class="n">fig_log2vsNormal</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Data-structure">Data structure<a class="anchor-link" href="#Data-structure"> </a></h1><p>Data like TPM is lognormal distributed.</p>
<ul>
<li>There cannot be a negative number of transcripts.</li>
<li><p>log transforms (like fold-change - log 2) are more normally disturbuted.</p>
<p>More irregular curves, like TP53 show more log-normal distributions when clustered into relevent groups, like tissue types.</p>
<p>Otherwise, something like sklearn.preprocessing.RobustScaler looks good for IQR scaling - robust to outliers.</p>
<p>This Scaler removes the median and scales the data according to the quantile range (defaults to IQR: Interquartile Range). The IQR is the range between the 1st quartile (25th quantile) and the 3rd quartile (75th quantile).</p>
</li>
</ul>
<h2 id="Correlation">Correlation<a class="anchor-link" href="#Correlation"> </a></h2><p>Gene expressions are correlated / anti-correlated (mutual information/effect) in interaction pathways.  'Activation' of certiain pathways is of interest for cancer and drug effects measures.</p>
<p>PCA with 'whitening'</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># Data is now effectively in fold change units</span>

<span class="c1"># Use a minum value to prevent div by zero.</span>
<span class="c1">#safelog2 = lambda x: numpy.log2(x) if x else numpy.log2(sys.float_info.min)</span>
<span class="n">safelog2</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
<span class="n">gfc</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="n">safelog2</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">gfc</span><span class="o">.</span><span class="n">info</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gfc</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">PowerTransformer</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">RobustScaler</span>

<span class="c1"># Which way are we scaling? Rows or columns?</span>
<span class="n">rs</span> <span class="o">=</span> <span class="n">RobustScaler</span><span class="p">()</span>
<span class="n">scaler</span> <span class="o">=</span> <span class="n">rs</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">gfc</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>
<span class="n">gfc_norm</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">scaler</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">gfc</span><span class="o">.</span><span class="n">T</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="n">gfc</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">gfc</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">gfc_norm</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gfc_norm</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="n">gfc_norm</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;PTEN&#39;</span>
<span class="n">data_lists</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">labels</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">tissue</span> <span class="ow">in</span> <span class="n">tissue_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">tissue_dict</span><span class="p">[</span><span class="n">tissue</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">]</span>
    <span class="c1"># print(tissue)</span>
    <span class="c1"># print(len(samples))</span>
    <span class="n">gt_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples</span> <span class="k">if</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">gfc_norm</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gt_samples</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">:</span>
        <span class="n">dt</span> <span class="o">=</span> <span class="n">gfc_norm</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="n">gt_samples</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">dt</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">data_lists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt</span><span class="p">)</span>
            <span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1">(</span><span class="si">{}</span><span class="s1">) ~ </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">tissue</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">samples</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">dt</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="mi">1</span><span class="p">)))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only </span><span class="si">{}</span><span class="s1"> samples for &quot;</span><span class="si">{}</span><span class="s1">&quot;&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gt_samples</span><span class="p">),</span> <span class="n">tissue</span><span class="p">))</span>
<span class="c1"># add all samples</span>
<span class="n">dt</span> <span class="o">=</span> <span class="n">gfc_norm</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="n">data_lists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dt</span><span class="p">)</span>
<span class="n">labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> - all (</span><span class="si">{}</span><span class="s1">) ~ </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">dt</span><span class="p">),</span> <span class="nb">round</span><span class="p">(</span><span class="n">dt</span><span class="o">.</span><span class="n">median</span><span class="p">(),</span> <span class="mi">1</span><span class="p">)))</span>

<span class="n">fig_log2tissue</span> <span class="o">=</span> <span class="n">ff</span><span class="o">.</span><span class="n">create_distplot</span><span class="p">(</span><span class="n">data_lists</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">show_hist</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_rug</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show_curve</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># curve_type=&#39;normal&#39;)</span>
<span class="n">fig_log2tissue</span><span class="o">.</span><span class="n">layout</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span>
<span class="n">fig_log2tissue</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">confusion_matrix</span><span class="p">,</span> <span class="n">accuracy_score</span>
<span class="kn">from</span> <span class="nn">sklearn.discriminant_analysis</span> <span class="kn">import</span> <span class="n">LinearDiscriminantAnalysis</span> <span class="k">as</span> <span class="n">LDA</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>

<span class="kn">import</span> <span class="nn">sklearn.decomposition</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sklearn.decomposition&#39;</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">sklearn</span><span class="o">.</span><span class="n">decomposition</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">f</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;_&#39;</span><span class="p">])</span>

<span class="kn">import</span> <span class="nn">sklearn.discriminant_analysis</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sklearn.discriminant_analysis&#39;</span><span class="p">)</span>
<span class="n">pprint</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">sklearn</span><span class="o">.</span><span class="n">discriminant_analysis</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">f</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;_&#39;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sklearn.ensemble&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">sklearn.ensemble</span>
<span class="n">pprint</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">sklearn</span><span class="o">.</span><span class="n">ensemble</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">f</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;_&#39;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;sklearn.metrics&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">sklearn.metrics</span>
<span class="n">pprint</span><span class="p">([</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">sklearn</span><span class="o">.</span><span class="n">metrics</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">f</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;_&#39;</span><span class="p">])</span>
<span class="n">pprint</span><span class="p">(</span><span class="nb">dir</span><span class="p">(</span><span class="n">sklearn</span><span class="o">.</span><span class="n">metrics</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1"># sparse PCA is probably better - interpret which genes are more important.</span>

<span class="n">pca3</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># Alt opt PCA(0.99) - enough components for 99% of variance</span>
<span class="n">sparse_pca</span> <span class="o">=</span> <span class="n">sklearn</span><span class="o">.</span><span class="n">decomposition</span><span class="o">.</span><span class="n">SparsePCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="n">sparse_comp</span> <span class="o">=</span> <span class="n">sparse_pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">gfc_norm</span><span class="p">)</span>
<span class="n">components3</span> <span class="o">=</span> <span class="n">pca3</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">gfc_norm</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">components3</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="n">pca3</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sparse_comp</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sparse_pca</span><span class="o">.</span><span class="n">components_</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">gfc_norm</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sparse_comp</span><span class="p">)</span>
<span class="n">pc1</span> <span class="o">=</span> <span class="n">sparse_comp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c1">#pc1_genes = [g.index for g in pc1 if g != 0]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span>
<span class="n">pc1</span> <span class="o">=</span> <span class="n">pc1</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">abs</span><span class="p">)</span>
<span class="c1">#print(pc1.sort_values.__doc__)</span>
<span class="n">pc1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span>
<span class="n">pc1_gene_list</span> <span class="o">=</span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">pc1</span><span class="o">.</span><span class="n">index</span> <span class="k">if</span> <span class="n">pc1</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pc1_gene_list</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gtm</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gtex_tpm_med_fn</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">gtm</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s1">&#39;Description&#39;</span><span class="p">,</span> <span class="s1">&#39;Name&#39;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1">#tissues = [(col.split(&#39; - &#39;)[0], col.replace(col.split(&#39; - &#39;)[0] + &#39; - &#39;, &#39;&#39;)) for col in gtm.columns]</span>
<span class="n">tissues</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; - &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">gtm</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
<span class="n">subtypes</span> <span class="o">=</span>  <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; - &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; - &#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">gtm</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

<span class="n">gtm</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">tissues</span><span class="p">,</span> <span class="n">subtypes</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gtm</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="nb">print</span><span class="p">(</span><span class="n">gtm</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;TP53&#39;</span><span class="p">,</span> <span class="s1">&#39;Brain&#39;</span><span class="p">])</span>

<span class="c1"># Makes some reasonable subtables.  With droplevel seems to work ok.</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gtm</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s1">&#39;Brain&#39;</span><span class="p">]</span>
<span class="c1">#gtm.loc[:, gtm.columns.get_level_values(1) == &#39;Cortex&#39;]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Full-size---slow-run">Full size - slow run<a class="anchor-link" href="#Full-size---slow-run"> </a></h1>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gtex_attr</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Full list of patient samples is </span><span class="si">{</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s2">&quot;SMAFRZE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1"># Use only the RNASEQ set</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using only RNASEQ drops </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s2">&quot;SMAFRZE&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;RNASEQ&quot;</span><span class="p">])</span><span class="si">}</span><span class="s1"> records&#39;</span><span class="p">)</span>
<span class="n">rna_seq_patients</span> <span class="o">=</span> <span class="n">gtex_attr</span><span class="p">[</span><span class="n">gtex_attr</span><span class="p">[</span><span class="s2">&quot;SMAFRZE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;RNASEQ&quot;</span><span class="p">][</span><span class="s2">&quot;SAMPID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>

<span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">patient_list</span> <span class="o">=</span> <span class="n">rna_seq_patients</span>
<span class="n">patient_list</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">rna_seq_patients</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rna_seq_patients</span><span class="p">)</span> <span class="o">/</span> <span class="mi">4</span><span class="p">))</span>

<span class="c1"># Leaves a reserve list for validating with later.</span>
<span class="n">reserve_patients</span> <span class="o">=</span> <span class="p">[</span><span class="n">patient</span> <span class="k">for</span> <span class="n">patient</span> <span class="ow">in</span> <span class="n">rna_seq_patients</span> <span class="k">if</span> <span class="n">patient</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">patient_list</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Total of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">patient_list</span><span class="p">)</span><span class="si">}</span><span class="s1"> patients&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">patient_list</span><span class="p">)</span><span class="si">}</span><span class="s1"> RNASEQ patients with a reserve of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">reserve_patients</span><span class="p">)</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting g8 load&quot;</span><span class="p">)</span>
<span class="n">_g8</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">gtex_tpm_fn</span><span class="p">,</span>
        <span class="n">skiprows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>  <span class="c1"># First 2 rows are comments of file size - other func here to select genes</span>
        <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Description&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">patient_list</span><span class="p">,</span>  <span class="c1"># Number of samples/patients limit</span>
    <span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">finished g8 load&quot;</span><span class="p">)</span>
<span class="c1"># g8 = _g8.drop(&#39;gene_id&#39;, axis=1)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">g8</span> <span class="o">=</span> <span class="n">_g8</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Description&#39;</span><span class="p">)</span>
<span class="n">g8</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">rename_cols</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1">#g8.head()</span>
<span class="nb">dir</span><span class="p">(</span><span class="n">g8</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">_g8</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Description&#39;</span><span class="p">],</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">g8</span> <span class="o">=</span> <span class="n">_g8</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Description&quot;</span><span class="p">)</span>
<span class="n">g8t</span> <span class="o">=</span> <span class="n">g8</span><span class="o">.</span><span class="n">T</span>
<span class="n">pca3</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># Alt opt PCA(0.99) - enough components for 99% of variance</span>
<span class="n">sparse_pca</span> <span class="o">=</span> <span class="n">sklearn</span><span class="o">.</span><span class="n">decomposition</span><span class="o">.</span><span class="n">SparsePCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;starting sparse_pca&quot;</span><span class="p">)</span>
<span class="n">sparse_comp</span> <span class="o">=</span> <span class="n">sparse_pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">g8t</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;starting pca&quot;</span><span class="p">)</span>
<span class="n">components3</span> <span class="o">=</span> <span class="n">pca3</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">g8t</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">components3</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="n">pca3</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sparse_comp</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">sparse_pca</span><span class="o">.</span><span class="n">components_</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">g8t</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="c1"># print(sparse_comp)</span>
<span class="n">pc1</span> <span class="o">=</span> <span class="n">sparse_comp</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c1">#pc1_genes = [g.index for g in pc1 if g != 0]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pc1</span><span class="p">)</span>
<span class="n">pc1a</span> <span class="o">=</span> <span class="n">pc1</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">abs</span><span class="p">)</span>
<span class="c1">#print(pc1.sort_values.__doc__)</span>
<span class="n">pc1a</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pc1a</span><span class="p">)</span>
<span class="n">pc1_gene_list</span> <span class="o">=</span><span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">pc1a</span><span class="o">.</span><span class="n">index</span> <span class="k">if</span> <span class="n">pc1a</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
<span class="c1">#print(pc1_gene_list)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pc1_gene_list</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">sig_genes</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">pc1a</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">pc1a</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">sig_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">sig_genes</span><span class="p">)</span>

<span class="n">pprint</span><span class="p">(</span><span class="n">sig_genes</span><span class="p">[:</span><span class="mi">50</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pheno_desc</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEX_PHENO_DD</span><span class="p">)))</span>
<span class="n">pheno_desc</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pheno</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">GTEX_PHENO_DS</span><span class="p">)),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">pheno</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gtex_attr</span><span class="p">[[</span><span class="s2">&quot;SAMPID&quot;</span><span class="p">,</span> <span class="s2">&quot;SMTS&quot;</span><span class="p">,</span> <span class="s2">&quot;SMTSD&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">tissue</span> <span class="o">=</span> <span class="n">gtex_attr</span><span class="p">[[</span><span class="s2">&quot;SAMPID&quot;</span><span class="p">,</span> <span class="s2">&quot;SMTS&quot;</span><span class="p">,</span> <span class="s2">&quot;SMTSD&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">tissue</span><span class="p">[</span><span class="s2">&quot;SMTS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tissue</span><span class="p">[</span><span class="s2">&quot;SMTS&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
<span class="n">tissue</span><span class="p">[</span><span class="s2">&quot;SMTSD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tissue</span><span class="p">[</span><span class="s2">&quot;SMTSD&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_df</span> <span class="o">=</span> <span class="n">tissue</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">g8t</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;SAMPID&quot;</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">fastai.tabular.all</span> <span class="kn">import</span> <span class="n">Categorify</span><span class="p">,</span> <span class="n">FillMissing</span><span class="p">,</span> <span class="n">TabularPandas</span><span class="p">,</span> <span class="n">cont_cat_split</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">cont</span><span class="p">,</span><span class="n">cat</span> <span class="o">=</span> <span class="n">cont_cat_split</span><span class="p">(</span><span class="n">train_df</span><span class="p">,</span> <span class="n">max_card</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">dep_var</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SMTS&quot;</span><span class="p">,</span> <span class="s2">&quot;SMTSD&quot;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">cont</span><span class="p">,</span><span class="n">cat</span> <span class="o">=</span> <span class="n">cont_cat_split</span><span class="p">(</span><span class="n">train_df</span><span class="p">,</span> <span class="n">max_card</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">dep_var</span><span class="o">=</span><span class="s2">&quot;SMTSD&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">train_ob</span> <span class="o">=</span> <span class="n">TabularPandas</span><span class="p">(</span><span class="n">train_df</span><span class="p">,</span> <span class="n">procs</span><span class="o">=</span><span class="p">[</span><span class="n">FillMissing</span><span class="p">,</span> <span class="n">Categorify</span><span class="p">],</span> <span class="n">cat_names</span><span class="o">=</span><span class="n">cat</span><span class="p">,</span> <span class="n">cont_names</span><span class="o">=</span><span class="n">cont</span><span class="p">,</span> <span class="n">y_names</span><span class="o">=</span><span class="s2">&quot;SMTSD&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">dir</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">sorted</span><span class="p">([(</span><span class="n">x</span><span class="p">,</span> <span class="n">sys</span><span class="o">.</span><span class="n">getsizeof</span><span class="p">(</span><span class="nb">globals</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1">#%memit</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">%</span><span class="k">memit</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

